pkg/pillar/cmd/zfsmanager/handlediskconfig.go

```diff
@@ -0,0 +1,228 @@
+					}
+				case info.StorageStatus_STORAGE_STATUS_OFFLINE:
+					continue
+				default:
+					log.Errorf("unexpected state of disk %s (%s) to make offline", devName, devStat.State.String())
+					continue
+				}
+			case types.EdgeNodeDiskConfigTypeUnused:
+				if stdout, err := zfs.RemoveVDev(log, vault.DefaultZpool, devName); err != nil {
+					log.Errorf("cannot remove %s: %s %s", devName, stdout, err)
+				}
+			}
+		}
+	}
+	for _, el := range disks.Children {
+		// process children states
+		// we process only states of devices as part of vdevs, so we assume that we can handle them in any order
+		disksStateProcessing(el, pool)
+	}
+}
+
+//disksLayoutProcessing iterate over disks and adjust pool layout accordingly to the config
+func disksLayoutProcessing(disks types.EdgeNodeDisks, pool libzfs.Pool) {
+	vdevTree, err := pool.VDevTree()
+	if err != nil {
+		log.Errorf("cannot get vdev tree: %s", err)
+		return
+	}
+	switch disks.ArrayType {
+	case types.EdgeNodeDiskArrayTypeRAID0:
+		disksLayoutRaid0Process(vdevTree, disks.Children)
+	default:
+		// TBD: other array types
+		log.Warnf("Not implemented layout processing for array type: %d", disks.ArrayType)
+	}
+}
+
+//disksLayoutRaid0Process ensure layout for batch of top-level vdevs in pool
+//we support raid0 of unspecified (single disk) and raid1 (mirror of disks) layout here
+func disksLayoutRaid0Process(vdevTree libzfs.VDevTree, disks []types.EdgeNodeDisks) {
+	for _, el := range disks {
+		switch el.ArrayType {
+		case types.EdgeNodeDiskArrayTypeUnspecified, types.EdgeNodeDiskArrayTypeRAID1:
+			diskName := ""
+			// check if we have one of defined devices as part of vdev
+			for _, dsk := range el.Disks {
+				if currentDiskName, _ := getVdevNameAndStatus(vdevTree, dsk.Disk.Name); currentDiskName != "" {
+					diskName = dsk.Disk.Name
+					break
+				}
+			}
+			// if no disk found add first as a new vdev
+			if diskName == "" {
+				if len(el.Disks) > 0 {
+					// we add first device here as a new vdev to attach needed disks to it later
+					if stdout, err := zfs.AddVDev(log, vault.DefaultZpool, el.Disks[0].Disk.Name); err != nil {
+						log.Errorf("cannot add %s: %s %s", diskName, stdout, err)
+					} else {
+						diskName = el.Disks[0].Disk.Name
+					}
+				}
+			}
+			// if disk is here attach another disks
+			if diskName != "" {
+				for _, dsk := range el.Disks {
+					// check if already in pool or added as part of current iteration (we do not refresh the tree)
+					if currentDiskName, _ := getVdevNameAndStatus(vdevTree, dsk.Disk.Name); currentDiskName != "" || diskName == dsk.Disk.Name {
+						continue
+					}
+					if stdout, err := zfs.AttachVDev(log, vault.DefaultZpool, diskName, dsk.Disk.Name); err != nil {
+						log.Errorf("cannot attach %s to %s: %s %s", dsk.Disk.Name, diskName, stdout, err)
+					}
+				}
+			}
+		default:
+			log.Warnf("No supported child processing for array type: %d", el.ArrayType)
```

Comment:
In a follow-up it might make sense to add some API to return errors i..e., a DiskStatus corresponding to the DiskConfig and make the Info protobuf API carry ErrorInfo.